indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java

```diff
@@ -0,0 +1,1166 @@
+    protected static final String SHOULD_COMMIT_METADATA_KEY = "io.druid.indexer.updater.converter.shouldCommitMetadata";
+    protected static final String PUBLISHED_SEGMENT_KEY = "io.druid.indexer.updater.converter.publishedSegment";
+    private static final Logger log = new Logger(ConvertingOutputFormat.class);
+
+    @Override
+    public RecordWriter<Text, Text> getRecordWriter(TaskAttemptContext context) throws IOException, InterruptedException
+    {
+      return new RecordWriter<Text, Text>()
+      {
+        @Override
+        public void write(Text key, Text value) throws IOException, InterruptedException
+        {
+          // NOOP
+        }
+
+        @Override
+        public void close(TaskAttemptContext context) throws IOException, InterruptedException
+        {
+          // NOOP
+        }
+      };
+    }
+
+    @Override
+    public void checkOutputSpecs(JobContext context) throws IOException, InterruptedException
+    {
+      // NOOP
+    }
+
+    @Override
+    public OutputCommitter getOutputCommitter(final TaskAttemptContext context)
+        throws IOException, InterruptedException
+    {
+      final MetadataStorageUpdaterJobHandler handler = getInjector(context.getConfiguration()).getInstance(
+          MetadataStorageUpdaterJobHandler.class
+      );
+
+      return new OutputCommitter()
+      {
+        @Override
+        public void setupJob(JobContext jobContext) throws IOException
+        {
+
+        }
+
+        @Override
+        public void setupTask(TaskAttemptContext taskContext) throws IOException
+        {
+
+        }
+
+        @Override
+        public boolean needsTaskCommit(TaskAttemptContext taskContext) throws IOException
+        {
+          return taskContext.getConfiguration().get(PUBLISHED_SEGMENT_KEY) != null;
+        }
+
+        @Override
+        public void commitTask(final TaskAttemptContext taskContext) throws IOException
+        {
+          final Progressable commitProgressable = new Progressable()
+          {
+            @Override
+            public void progress()
+            {
+              taskContext.progress();
+            }
+          };
+          final HadoopDruidConverterConfig config = converterConfigFromConfiguration(taskContext.getConfiguration());
+          final String finalSegmentString = taskContext.getConfiguration().get(PUBLISHED_SEGMENT_KEY);
+          if (finalSegmentString == null) {
+            throw new IOException("Could not read final segment");
+          }
+          final DataSegment newSegment = HadoopDruidConverterConfig.jsonMapper.readValue(
+              finalSegmentString,
+              DataSegment.class
+          );
+          log.info("Committing new segment");
+          taskContext.progress();
+          // Usually the indexing service handles its own commits, but standalone hadoop does not
+          if (shouldCommitMetadata(taskContext.getConfiguration())) {
```

Comment:
I'm fairly certain this will attempt to write metadata as part of the reducer, which is actually not good.  A reducer can succeed and the total job fail, in which case we don't want the segment in the DB at all (it will be loaded by a historical node and just sit there, not able to do anything because its partitions weren't completed).  The segment metadata should be written all together in a single transaction to the DB once we know that the job has completed successfully.